segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/tables/TableEntry.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.contracts.tables;
+
+import io.pravega.common.util.ArrayView;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * An Entry in a Table Segment, made up of a Key and a Value, with optional Version.
+ */
+@Getter
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+public class TableEntry {
+    /**
+     * The Key.
+     */
+    private final TableKey key;
+
+    /**
+     * The Value (data) of the entry.
+     */
+    private final ArrayView value;
+    /**
+     * Creates a new instance of the TableEntry class with no desired version.
+     *
+     * @param key   The Key.
+     * @param value The Value.
+     */
+    public static TableEntry unversioned(@NonNull ArrayView key, @NonNull ArrayView value) {
```

Comment:
Shouldn't table entries be always versioned? If I make a table entry unversioned, then what does it mean to perform a conditional update to the corresponding key-value? For example, in ZK, znodes are always versioned, and unconditional updates will simply ignore the version.

Reply:
There is no separate API for conditional vs unconditional updates (check `TableSegment`). Whether `TableEntry` has a version or not determines whether that particular entry update is versioned or not.

Reply:
Ok, so if a table entry has a version, then all updates are conditional on the version? we won't accept unconditional updates to a versioned entry?

Reply:
To be more specific, I was checking this signature and the associated javadoc:

https://github.com/pravega/pravega/pull/2955/files#diff-d599f95fe5b91b0fd6be8c9e124fa22bR50

I think the semantics is the following:
* If the table entry provided as an argument has a version, then the put is performed conditionally. If this is right, what happens in the case the current key in the table has no version associated?
* If the table entry provided as an argument has no version, then the put is performed conditionally.

Reply:
Yes. `TableEntry` here does not mean a physical entry in the table. This means that I want to update the entry with the given key, given value and (optionally) condition it on a specific version. 

I could have made this `TableEntryUpdate` to disambiguate but I return the exact same object when I invoke `get` (it returns a `TableEntry` which has key, value and version). 

And yes, to clarify (I hope that the Javadoc you pointed to would have been clear):
- If at least one TableEntry from the List of TableEntry instances provided to `put` has a non-null version, then the entire List (batch) will be conditioned on those TableEntry instances that have versions. The entire List will either go in at once or be fully rejected based on the outcome of the condition check.
- If none of the TableEntry instances in the List have explicit versions, then this is an unconditional update, and the batch will go through (again, atomically).